source/extensions/outlier_detection_monitors/common/monitor_base_impl.cc

```diff
@@ -0,0 +1,79 @@
+#include "source/extensions/outlier_detection_monitors/common/monitor_base_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Outlier {
+
+bool HTTPCodesBucket::match(const ExtResult& result) const {
+  // We should not get here with errors other then HTTP codes.
+  ASSERT(matchType(result));
+  const HttpCode& http_code = absl::get<HttpCode>(result);
+  return ((http_code.code() >= start_) && (http_code.code() <= end_));
+}
+
+bool LocalOriginEventsBucket::match(const ExtResult& event) const {
+  // We should not get here with errors other then Local Origin events.
+  ASSERT(matchType(event));
+  const LocalOriginEvent& local_origin_event = absl::get<LocalOriginEvent>(event);
+  // Capture all events except the success
+  return (!((local_origin_event.result() == Result::LocalOriginConnectSuccessFinal) ||
+            (local_origin_event.result() == Result::ExtOriginRequestSuccess)));
+}
+
+void ExtMonitorBase::putResult(const ExtResult& result) {
+  if (buckets_.empty()) {
+    return;
+  }
+
+  bool matchedError = false;
+  bool matchedType = false;
+  // iterate over all error buckets
+  for (auto& bucket : buckets_) {
+    // if the bucket is not interested in this type of result/error
+    // just ignore it.
+    if (!bucket->matchType(result)) {
+      continue;
+    }
+
+    matchedType = true;
+
+    // check if the bucket "catches" the result.
+    if (bucket->match(result)) {
+      matchedError = true;
+      break;
+    }
+  }
+
+  // If none of buckets had the matching type, just bail out.
+  if (!matchedType) {
+    return;
+  }
+
+  if (matchedError) {
+    // Count as error.
+    if (onError()) {
+      callback_(enforce_, name(), absl::nullopt /* extra_info */);
+      // Reaching error was reported via callback.
+      // but the host may or may not be ejected based on enforce_ parameter.
+      // Reset the monitor's state, so a single new error does not
+      // immediately trigger error condition again.
+      onReset();
+    }
```

Comment:
Is this thread safe?

Reply:
The callback is not and as I indicated above (in one of the comments) I will try to rework it. `onError` and `onReset` must be thread safe. They usually use atomics to manipulate counters.

Reply:
Refactored the callback it to be thread safe.